# 組み込み関数について

---

- ref: https://docs.python.org/ja/3.7/library/functions.html
- blog: https://slash-z.com/
- github: https://github.com/KazutoMakino/PythonCourses

注意：全てについて説明はせず，良く使われる処理のみについて記載いたします

---

## 0. 目次: <a class="anchor" id="0"></a>

1. [組み込み関数について](#1)

---

---

## 1. 組み込み関数について <a class="anchor" id="1"></a>  
[目次に戻る](#0)  

- `abs(x)`: `x` の絶対値を返します．引数が複素数である場合はノルム（複素平面上における複素ベクトルの長さ）を返します．

In [1]:
abs(5), abs(-5), abs(1+1j)

(5, 5, 1.4142135623730951)

最後の出力は $\sqrt{2} = 1.41421356...$ です．

- `all(iterable)`: `iterable` の全ての要素が真，もしくは，空であれば `True` を返します．

In [15]:
(
    all([True]),
    all((0,1)),
    all([True is True, False is False, True is not False])
)

(True, False, True)

- `any(iterable)`: `iterable` のいずれかの要素が真ならば `True` を返します．`iterable` が空なら `False` を返します．

In [16]:
(
    any([True]),
    any((0,1)),
    any([True is True, False is False, True is not False])
)

(True, True, True)

- `bool(x)`: ブール値を返します．`x` が偽または省略されている場合は `False` を返し，それ以外では `True` を返します．

In [24]:
(
    bool(1),
    bool(0),
    bool(-100),
    bool(100),
    bool(1+1j),
    bool(""),
    bool(" ")
)

(True, False, True, True, True, False, True)

`False` は以下の真理値判定手続きに基づくとのことで，これら以外は `True` (https://docs.python.org/ja/3.7/library/stdtypes.html#truth) ．

- 偽であると定義されている定数
- 数値型におけるゼロ
- 空のシーケンスまたはコレクション

- `complex(real, imag)`: 値 `real + imag*j` の複素数を返すか，文字列や数を複素数に変換します．第一引数が文字列なら，それが複素数と解釈され，この関数は第二引数なしで呼び出さなければなりません．第二引数は文字列であってはなりません．それぞれの引数は（複素数を含む）任意の数値型です．`imag` が省略された場合，標準の値はゼロで，このコンストラクタは `int` や `float` のような数値変換として働きます．両方の引数が省略された場合，`0j` を返します．引数に文字列を取る場合は，この文字列が複素数型ライクにする必要があり，かつ，実部と虚部の間の符号の前後に半角スペースを入れると `TypeError`になるので注意．

In [34]:
complex(1, 1), complex("1+1j")

((1+1j), (1+1j))

In [35]:
complex("1 + 1j")

ValueError: complex() arg is a malformed string

- `dict(**kwarg)` または `dict(mapping, **kwarg)` または `dict(iterable, **kwarg)`: 新しい辞書を作成します．

In [38]:
dict(map("1",1))

TypeError: 'int' object is not iterable

- `dict(**kwarg)` または `dict(mapping, **kwarg)` または `dict(iterable, **kwarg)`: 新しい辞書を作成します．  
1 つ目は，`dict(key=value)` とすることにより辞書を作成します．  
2 つ目は，`dict(zip(key_list, value_list))` とすることにより辞書を作成します．  
3 つ目は，`dict([(key, value]))` とすることにより辞書を作成します．  

In [6]:
# dict(**kwarg)
dict(a=0, b=1, c=2)

{'a': 0, 'b': 1, 'c': 2}

In [3]:
# dict(mapping, **kwarg)
k = ["a", "b", "c"]
v = [0, 1, 2]
dict(zip(k, v))

{'a': 0, 'b': 1, 'c': 2}

In [5]:
# dict(iterable, **kwarg)
dict([("a", 0), ("b", 1), ("c", 2)])

{'a': 0, 'b': 1, 'c': 2}

- `divmod(a, b)`: 複素数でない 2 つの数を引数として取り，整数の除去を行った時の商と余りからなるペアを返します．この返り値は，`(a // b, a % b)` と同じです．

In [7]:
divmod(5, 3)

(1, 2)

In [9]:
divmod(2**16 - 1, 17)

(3855, 0)

- `enumerate(iterable, start=0)`: enumerate オブジェクトを返します．リスト化には，list 関数を用います．

In [11]:
a = ["a", "b", "c"]
b = enumerate(a)
b

例えばリスト化すると，インデックスと要素のペアをタプルで返します．

In [12]:
list(b)

[(0, 'a'), (1, 'b'), (2, 'c')]

- `eval(expression, globals, locals)`: 文字列とオプションの引数 `globals`, `locals` を取ります．オプション引数の `globals` は辞書で，`locals` はマッピングオブジェクトでなければなりません．

In [13]:
eval("1+2+3")

6

In [14]:
x = "hello"
y = "world"
eval("x + ' ' + y")

'hello world'

`eval` を用いると，このように文字列を式のように処理することができます．

- `exec(object, globals, locals)`: この関数は python コードの動的な実行をサポートします．`object` は文字列化コードオブジェクトでなければなりません．オプション引数の `globals` は辞書で，`locals` はマッピングオブジェクトでなければなりません．

In [20]:
exec("a=1")
a

1

In [22]:
x = "hello"
y = "world"
exec("z = x + ' ' + y")
z

'hello world'

- `float(x)`: 数または文字列 `x` から生成された浮動小数点数を返します．引数が文字列の場合，10進数を含んだ文字列にしてください．

In [ ]:
float(1), float("-1"), float(1+2), float()

- `help(object)`: `object` のヘルプを表示します．引数が与えられていない場合，対話的ヘルプシステムはインタプリタコンソール上で起動します．

In [4]:
help(abs)

Help on built-in function abs in module builtins:

abs(x, /)
    Return the absolute value of the argument.



In [9]:
a = None
help(a)

Help on NoneType object:

class NoneType(object)
 |  Methods defined here:
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



- `input(prompt)`: 引数 `prompt` が存在すれば，それが末尾の開業を除いて標準出力に書き出されます．次に，この関数は入力から1行を読み込み，文字列に変換して（末尾の開業を除いて）返します．EOF が読み込まれたとき，`EOFError` が送出されます．

In [12]:
input()

 aaaaaaa


'aaaaaaa'

In [13]:
input("hello ")

hello  world


'world'

- `int(x)` または `int(x, base=10)`: 数値または文字列 `x` から生成された整数オブジェクトを返します．引数が与えられない場合には 0 を返します．

In [21]:
int(1), int(-1), int("1"), int(0.999), int(float(1)), int()

(1, -1, 1, 0, 1, 0)

- `isinstance(object, classinfo)`: `object` 引数が `classinfo` 引数のインスタンスであるか，サブクラスのインスタンスの場合に `True` を返します．`object` が与えられた方のオブジェクトでない場合，この関数は常に `False` を返します．`classinfo` が型オブジェクトのタプルの場合，`object` がそれらのいずれかのインスタンスであれば `True` を返します．`classinfo` が型や方からなるタプル，あるいは複数のタプルのいずれでもない場合，`TypeError` が送出されます．

In [26]:
a = "test"
b = 1
isinstance(a, str), isinstance(b, int), isinstance(b, float), isinstance(b, (str, int, float))

(True, True, False, True)

- `len(s)`: オブジェクト `s` の長さ（要素の数）を返します．引数はシーケンスかコレクションです．

In [29]:
a = "test"
len(a)

4

In [30]:
a = [0, 1, 2, 3]
len(a)

4

In [33]:
a = {"a": 0, "b": 1}
len(a)

2

In [34]:
a = [[0, 1, 2, 3]]
len(a)

1

- `list(iterable)`: `iterable` のリストを返します．

In [35]:
list([0, 1])

[0, 1]

In [36]:
list((0, 1))

[0, 1]

In [37]:
list(enumerate(["a", "b", "c"]))

[(0, 'a'), (1, 'b'), (2, 'c')]

- `map(function, iterable, ...)`: `function` を，結果を返しながら `iterable` の全ての要素に適用するイテレータを返します．追加の `iterable` 引数が渡された場合は，`function` はその数だけの引数を取らなければならず，全てのイテラブルから並行して撮られた要素に適用されます．

In [45]:
a = ["1", "2", "3"]
b = map(int, a)
b

In [47]:
list(b)

[1, 2, 3]

In [48]:
list(b)

[]

上記 2 個目の `list(b)` が `[]` となったのは，`b` がイテレータであり，1 回目の `list(b)` でインデックスが最後まで送られ，要素が全て取り出されたためです．予めリストとして違う文字変数に代入しておくことで，これは回避できます．

In [49]:
a = ["1", "2", "3"]
b = list(map(int, a))
b

[1, 2, 3]

In [50]:
b

[1, 2, 3]

In [51]:
b

[1, 2, 3]

- `max(iterable, key, default)` または `max(arg1, arg2, *args, key)`: `iterable` の中で最大の要素，または 2 つ以上の引数の中で最大のものを返します．

In [52]:
a = [0, 1, 2]
max(a)

2

In [53]:
max(0, 1, 2)

2

In [57]:
max(True, False, -1)

True

In [54]:
max("6", "1", 1)

TypeError: '>' not supported between instances of 'int' and 'str'

In [56]:
max(1, None)

TypeError: '>' not supported between instances of 'NoneType' and 'int'

大小関係が比較できない場合は，`TypeError` が送出されます．  
`None` はゼロでなく空なので，数値と比較することができません．

- `min(iterable, key, default)` または `min(arg1, arg2, *args, key)`: `iterable` の中で最小の要素，または 2 つ以上の引数の中で最小のものを返します．性質や使い方は `max()` と同様です．

- `next(iterator, default)`: `iterator` の `__next__()` メソッドと呼び出すことにより，次の要素を取得します．イテレータが尽きている場合，`default` が与えられていればそれが返され，そうでなければ `StopIeration` が送出されます．

In [58]:
a = ["1", "2", "3"]
b = map(int, a)
b

In [59]:
next(b)

1

In [60]:
next(b)

2

In [61]:
next(b)

3

In [62]:
next(b)

StopIteration: 

In [63]:
a = ["1", "2", "3"]
b = map(int, a)
next(b, "end")

1

In [64]:
next(b, "end")

2

In [65]:
next(b, "end")

3

In [66]:
next(b, "end")

'end'

In [67]:
next(b, "end")

'end'

- `open(file, mode="r", buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)`: file を開き，対応するファイルオブジェクトを返します．ファイルを開くことができなければ， `OSError` が送出されます．  
`file` は path-like object （パスとして見なされるオブジェクト）で，開くファイルの (絶対または現在のワーキングディレクトリに対する相対) パス名を与えるものです．  
mode はオプションの文字列で，ファイルが開かれるモードを指定します．デフォルトは `'r'` で、読み込み用にテキストモードで開くという意味です．指定可能なモードは次の表の通りです。

| 文字 | 意味 |
| ---- | ---- |
| 'r'  | 読み込み用に開く (デフォルト) |
| 'w'  | 書き込み用に開き、まずファイルを切り詰める |
| 'x'  | 排他的な生成に開き、ファイルが存在する場合は失敗する |
| 'a'  | 書き込み用に開き、ファイルが存在する場合は末尾に追記する |
| 'b'  | バイナリモード |
| 't'  | テキストモード (デフォルト) |
| '+'  | ディスクファイルを更新用に開く (読み込み／書き込み) |

テキストモードでは，`encoding` が指定されていない場合に使われるエンコーディングはプラットフォームに依存します．  
デフォルトのモードは 'r' (開いてテキストの読み込み、'rt' と同義) です．バイナリの読み書きアクセスについては，モード 'w+b' はファイルを開いて 0 バイトに切り詰めます．'r+b' はファイルを切り詰めずに開きます．  
Python はバイナリとテキストの I/O を区別します．(mode 引数に 'b' を含めて) バイナリモードで開かれたファイルは，内容をいかなるデコーディングもせずに bytes オブジェクトとして返します．(デフォルトや， mode 引数に 't' が含まれたときの) テキストモードでは，ファイルの内容は str として返され，バイト列はまず，プラットフォーム依存のエンコーディングか，`encoding` が指定された場合は指定されたエンコーディングを使ってデコードされます．